README.md
Palemoon-Preloader
==================

The firefox preloader, adapted for the palemoon fork.
Original source here: http://sourceforge.net/projects/ffpreloader/
All credit goes to sixxgate for this preloader. My git repo just contains modifications that allows it to be compiled in VS2013, update some depreciated C++ syntax, and include binaries for the palemoon fork of firefox.

Planned features:
 - Manually specify the exe name, so one build is not limited to specific firefox forks.

Known Bugs
 - The preloader cannot kill or "unload" the palemoon process as it is. I don't think I'm going to fix this, I'll just remove all options related to "unloading palemoon". The consequence of this is, if you exit the preloader while palemoon is not running, the preloaded palemoon process will remain running on your system. You can kill it, or let it be- it won't affect your saved browsing session (if you have one).
 - If you start the preloader while palemoon is running, the preloader will try and fail to preload palemoon, resulting in a window opening every 10 seconds or so. This can be avoided by not starting the preloader while your browser is already running. If it's happening, exit your browser and start it again.

How it works
 - Whenever it detects that palemoon.exe is not running, it creates starts palemoon, with no extensions or tabs loaded, in a hidden window. This speeds up the start up time considerably if you don't have that many extensions and tabs. 


RegistryClass/ChangeLog.txt
Copyright (C)2004-2005 6XGate Incorporated

RegistryClass is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published
by the Free Software Foundation; either version 2.1 of the License, or
(at your option) any later version.

RegistryClass is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with RegistryClass; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

==========================================================================
Change Log
==========================================================================
v1.0 build 1040
* Added HKEY_PERFORMANCE_DATA hive to the main class

v1.0 build 1043
* Removed unused library links (now runs on non-IE4 systems)
* Code cleanup and released

v1.0 build 1046
* Added overridible functions to open / create keys with different access permissions

v1.0 build 1049
* Fixed a bug that caused an inaccurate ERROR_SUCCESS return on GetValue functions

RegistryClass/Helpers.cpp
// Helpers.cpp: Unexported helper functions
//////////////////////////////////////////////////////////////////////
//
//   Copyright (C)2004 6XGate Incorporated
//
//   This file is part of RegistryClass.
//
//   RegistryClass is free software; you can redistribute it and/or modify
//   it under the terms of the GNU Lesser General Public License as published
//   by the Free Software Foundation; either version 2.1 of the License, or
//   (at your option) any later version.
//
//   RegistryClass is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU Lesser General Public License for more details.
//
//   You should have received a copy of the GNU Lesser General Public License
//   along with RegistryClass; if not, write to the Free Software
//   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
/////////////////////////////////////////////////////////////////////////////

#include "stdafx.h"

#ifdef _DEBUG
#undef THIS_FILE
static char THIS_FILE[]=__FILE__;
#define new DEBUG_NEW
#endif

//////////////////////////////////////////////////////////////////////
// Extra Non Exported Helper Functions
//////////////////////////////////////////////////////////////////////

//////////////////////////////////////////////////////////////////////
//
DWORD RegRecurDeleteKey (HKEY hOpenKey, LPCTSTR pszSubKey) {
	HKEY hKey;
	DWORD nKeyCount, nKeyNameLen, nLen; LPTSTR pszKeyNameBuf = NULL; DWORD dwRet; ULONG i;

	// Make sure we don't have a NULL SubKey string
	if (pszSubKey && lstrlen(pszSubKey)) {
		if ((dwRet = RegOpenKeyEx(hOpenKey, pszSubKey, 0, KEY_ALL_ACCESS, &hKey))==ERROR_SUCCESS) {
			// Get information about the keys, and allocate a buffer for the names
			RegQueryInfoKey(hKey, NULL, NULL, NULL, &nKeyCount, &nKeyNameLen, NULL, NULL, NULL, NULL, NULL, NULL);
			if (nKeyCount > 0) {
				nKeyNameLen++;
				pszKeyNameBuf = (LPTSTR)malloc(nKeyNameLen * sizeof(TCHAR));
				for (i = 0; i < nKeyCount; i++) {
					nLen = nKeyNameLen;
					dwRet = RegEnumKeyEx(hKey, 0, pszKeyNameBuf, &nLen, NULL, NULL, NULL, NULL);
					if (dwRet==ERROR_NO_MORE_ITEMS) break;
					else if (dwRet!=ERROR_SUCCESS) { free(pszKeyNameBuf); return dwRet; }
					dwRet = RegRecurDeleteKey(hKey, pszKeyNameBuf);
					if (dwRet!=ERROR_SUCCESS) { free(pszKeyNameBuf); return dwRet; }
				}
				free(pszKeyNameBuf);
				pszKeyNameBuf = NULL;
			}
			RegCloseKey(hKey);
			RegDeleteKey(hOpenKey, pszSubKey);
		} else return dwRet;
	} else return ERROR_BADKEY;
	return ERROR_SUCCESS;
}

//////////////////////////////////////////////////////////////////////
// Copies a registry key all its subkeys and values to a destination
// key.
// Parameters:
//		hSrcKey:		The handle to an already opened key to copy
//		hDestKey:		The handle to an already opened key to copy to
BOOL RegCopyKey (HKEY hSrcKey, HKEY hDestKey) {
	ULONG c, nValueCount, nKeyCount, nValueNameLen, nKeyNameLen, nValueLen, nLen, nDLen;
	LPTSTR pszNameBuf = NULL;
	DWORD nRet, dwType = 0;
	HKEY hFromKey, hToKey;
	BYTE *pValueBuf = NULL;

	// Get the information from the key we need to copy it.
	nRet = ::RegQueryInfoKey(hSrcKey, NULL, NULL, NULL, &nKeyCount, &nKeyNameLen, NULL, &nValueCount, &nValueNameLen, &nValueLen, NULL, NULL);
	if (nRet!=ERROR_SUCCESS) return FALSE;

	// Make sure there is room for the null terminator character
	nKeyNameLen++; nValueNameLen++;

	// Alloc a name buffer
	pszNameBuf = (LPTSTR)malloc(nKeyNameLen);

	if (nKeyCount) for (c = 0; c < nKeyCount; c++) {
		// Get the key under the current key
		nLen = nKeyNameLen;
		nRet = ::RegEnumKeyEx(hSrcKey, c, pszNameBuf, &nLen, NULL, NULL, NULL, NULL);
		if (nRet!=ERROR_SUCCESS) { free(pszNameBuf); return FALSE; }

		// Open the source key
		nRet = ::RegOpenKeyEx(hSrcKey, pszNameBuf, 0, KEY_ALL_ACCESS, &hFromKey);
		if (nRet!=ERROR_SUCCESS) { free(pszNameBuf); return FALSE; }

		// Create a key with its name at the destination
		nRet = ::RegCreateKeyEx(hDestKey, pszNameBuf, NULL, NULL, NULL, KEY_ALL_ACCESS, NULL, &hToKey, NULL);
		if (nRet!=ERROR_SUCCESS) { free(pszNameBuf); return FALSE; }

		// Copy it and its keys
		nRet = ::RegCopyKey (hFromKey, hToKey);
		if (nRet!=TRUE) { free(pszNameBuf); return FALSE; }
		::RegCloseKey(hFromKey);
		::RegCloseKey(hToKey);
	}
	free(pszNameBuf);
	pszNameBuf = NULL;

	// Alloc a value buffer
	pValueBuf = (BYTE*)malloc(nValueLen);
	pszNameBuf = (LPTSTR)malloc(nValueNameLen);
	if (nValueCount) for (c = 0; c < nValueCount; c++) {
		// Now get the current value
		nLen = nValueNameLen;
		nDLen = nValueLen;
		nRet = ::RegEnumValue(hSrcKey, c, pszNameBuf, &nLen, NULL, &dwType, pValueBuf, &nDLen);
		if (nRet!=ERROR_SUCCESS) { free(pValueBuf); free(pszNameBuf); return FALSE; }

		// Copy the value to the new key
		nRet = ::RegSetValueEx(hDestKey, pszNameBuf, NULL, dwType, pValueBuf, nDLen);
		if (nRet!=ERROR_SUCCESS) { free(pValueBuf); free(pszNameBuf); return FALSE; }
	}
	free(pValueBuf);
	free(pszNameBuf);
	pValueBuf = NULL; pszNameBuf = NULL;
	return TRUE;
}